File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        dp = [-1] * len(nums)
        def dfs(idx):
            if idx >= len(nums): return 0
            if dp[idx] != -1: return dp[idx]
            dp[idx] = max(dfs(idx + 2), dfs(idx + 3)) + nums[idx]
            return dp[idx]
        return max(dfs(0), dfs(1))

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 1) return nums[0];
        if (n < 3) return Math.max(nums[0], nums[1]);
        int a = nums[n - 2], b = nums[n - 1], c = 0;
        for (int i = n - 3; i >= 0; i--) {
            int d = Math.max(b, c) + nums[i];
            c = b;
            b = a;
            a = d;
        }
        return Math.max(a, b);
    }
}